squirrel.frontier/src/main/java/org/dice_research/squirrel/queue/scorebasedfilter/IUriKeywiseFilter.java

```diff
@@ -0,0 +1,22 @@
+package org.dice_research.squirrel.queue.scorebasedfilter;
+
+import org.dice_research.squirrel.data.uri.CrawleableUri;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Interface for filtering URIs based on duplicity score
+ */
+public interface IUriKeywiseFilter<T> {
+
+    /**
+     * This method returns the {@link CrawleableUri}s to be added to the queue after filtering out the ones based on their score
+     *
+     * @param keyWiseUris            map of based on key {@link CrawleableUri}s to be filtered
+     * @param minNumberOfUrisToCheck minimum number of {@link CrawleableUri}s to be checked for their score before filtering
+     * @param criticalScore          the critical score to check for filtering
+     * @return {@link Map} of {@link CrawleableUri}s to be added to the queue with their scores
+     */
+    Map<CrawleableUri, Float> filterUrisKeywise(Map<T, List<CrawleableUri>> keyWiseUris, int minNumberOfUrisToCheck, float criticalScore);
```

Comment:
I think the signature of the method is not a good choice. It is focused very much on a simple and direct way to implement your score-related features but misses to structure the functionality in a way that enables a future usage of this interface. Apart from that, it creates issues in the methods that use it (as described in one of the queue classes).
* `keyWiseUris`: sure, you want to have this input so that is fine.
* `minNumberOfUrisToCheck`: is this really something that needs to be given as a parameter of this method? You could also define it as an attribute of the filter implementation since not all filter implementations may want to make use of it.
* `criticalScore`: to me, it looks like this is definitely a value that should be an attribute of the filter implementation. It is a classic threshold that is defined once and never changed throughout the different calls of the method.
* Return type `Map<CrawleableUri, Float>`. I think that this choice leads to a lot of issues. The classes that implement this interface are grouping the URIs by a key. But the return type forces them to remove the grouping. Instead, URIs are mapped to a score that 1) may not be possible by other, future implementations of this interface (simply because they may not use a score at all) and 2) could be solved in a much easier way by assigning the score to the single URI objects using the `addData` method.